signingworker/worker.py

```diff
@@ -0,0 +1,187 @@
+                task_id, run_id,
+                {"workerGroup": self.worker_type, "workerId": self.worker_type}
+            )
+            task = self.tc_queue.task(task_id)
+            task_graph_id = task["taskGroupId"]
+            validate_task(task, self.supported_signing_scopes)
+            self.sign(task_id, run_id, task)
+            log.debug("Completing: %s, r: %s", task_id, run_id)
+            self.tc_queue.reportCompleted(task_id, run_id)
+            log.debug("Complete: %s, r: %s, tg: %s", task_id, run_id,
+                      task_graph_id)
+        except taskcluster.exceptions.TaskclusterRestFailure as e:
+            log.exception("TC REST failure, %s", e.status_code)
+            if e.status_code == 409:
+                log.debug("Task already claimed, acking...")
+            else:
+                raise
+        except (TaskVerificationError, ValidationError):
+            log.exception("Cannot verify task, %s", body)
+            self.tc_queue.reportException(
+                task_id, run_id, {"reason": "malformed-payload"})
+        except Exception:
+            log.exception("Error processing %s", body)
+
+        message.ack()
+
+    @redo.retriable(attempts=10, sleeptime=5, max_sleeptime=30)
+    def get_manifest(self, url):
+        r = requests.get(url)
+        r.raise_for_status()
+        return r.json()
+
+    def sign(self, task_id, run_id, task):
+        payload = task["payload"]
+        manifest_url = payload["signingManifest"]
+        signing_manifest = self.get_manifest(manifest_url)
+        # TODO: better way to extract filename
+        url_prefix = "/".join(manifest_url.split("/")[:-1])
+        for e in signing_manifest:
+            # TODO: "mar" is too specific, change the manifest
+            file_url = "{}/{}".format(url_prefix, e["mar"])
+            abs_filename = self.download_and_sign_file(
+                task_id, run_id, file_url, e["hash"])
+            # Update manifest data with new values
+            e["hash"] = get_hash(abs_filename)
+            e["size"] = os.path.getsize(abs_filename)
+        _, manifest_file = tempfile.mkstemp()
+        with open(manifest_file, "wb") as f:
+            json.dump(signing_manifest, f, indent=2, sort_keys=True)
+        log.debug("Uploading manifest for t: %s, r: %s", task_id, run_id)
+        self.create_artifact(task_id, run_id, "public/env/manifest.json",
+                             manifest_file, "application/json")
+
+    def download_and_sign_file(self, task_id, run_id, url, checksum):
+        work_dir = tempfile.mkdtemp()
+        # TODO: better parsing
+        filename = urlparse.urlsplit(url).path.split("/")[-1]
+        abs_filename = os.path.join(work_dir, filename)
+        r = requests.get(url)
+        r.raise_for_status()
+        with open(abs_filename, 'wb') as fd:
+            for chunk in r.iter_content(4096):
+                fd.write(chunk)
+        got_checksum = get_hash(abs_filename)
```

Comment:
It's a minor thing, but if you reworked how you're calculating hashes, you could do while you're writing the file rather than rereading the thing you just wrote.